AlphaWallet/TokenScriptClient/Models/XMLHandler.swift

```diff
@@ -829,38 +812,79 @@ extension XMLHandler {
         return attributeTypeElement.at_xpath("origins/token-id".addToXPath(namespacePrefix: xmlContext.namespacePrefix), namespaces: xmlContext.namespaces)
     }
 
-    static func getOriginFunctionElement(fromAttributeTypeElement attributeTypeElement: XMLElement, xmlContext: XmlContext) -> XMLElement? {
+    static func getEthereumOriginElement(fromAttributeTypeElement attributeTypeElement: XMLElement, xmlContext: XmlContext) -> XMLElement? {
         return attributeTypeElement.at_xpath("origins/ethereum".addToXPath(namespacePrefix: xmlContext.namespacePrefix), namespaces: xmlContext.namespaces)
     }
 
     static func getOriginUserEntryElement(fromAttributeTypeElement attributeTypeElement: XMLElement, xmlContext: XmlContext) -> XMLElement? {
         return attributeTypeElement.at_xpath("origins/user-entry".addToXPath(namespacePrefix: xmlContext.namespacePrefix), namespaces: xmlContext.namespaces)
     }
 
+    static func getEventParameterName(fromEthereumEventElement ethereumEventElement: XMLElement) -> String? {
+        guard let eventParameterName = ethereumEventElement["select"] else { return nil }
+        return eventParameterName
+    }
+
+    static func getEventDefinition(fromContractElement contractElement: XMLElement, xmlContext: XmlContext) -> EventDefinition? {
+        let addressElements = XMLHandler.getAddressElements(fromContractElement: contractElement, xmlContext: xmlContext)
+        guard let address = addressElements.first?.text.flatMap({ AlphaWallet.Address(string: $0.trimmed)}) else { return nil }
+        guard let eventName = contractElement.at_xpath("asnx:module", namespaces: xmlContext.namespaces)?["name"] else { return nil }
```

Comment:
Is the attribute ("name") renamed to "module" now?

```
    <xs:element name="event">
        <xs:complexType>
            <!-- TODO: this should be whatever ASN.X allows as element name -->
            <xs:attribute name="module" type="xs:NCName"/>
            <xs:attribute name="filter" type="xs:string"/>
            <xs:attribute name="select" type="xs:NCName"/>
        </xs:complexType>
    </xs:element>
```

Reply:
<ts:contract name="ETHRegistrarController">
    <ts:address network="1">0x283af0b28c62c092c9727f1ee09c02ca627eb7f5</ts:address>
    <asnx:module name="NameRegistered">
      <sequence>
        <element name="name" ethereum:type="string"/>
        <element name="label" ethereum:type="bytes32" ethereum:indexed="true"/>
        <element name="owner" ethereum:type="address" ethereum:indexed="true"/>
        <element name="cost" ethereum:type="uint"/>
        <element name="expires" ethereum:type="uint"/>
      </sequence>
    </asnx:module>
  </ts:contract>

Reply:
> Has the attribute ("name") renamed to "module" now?

No. `moulde="xxx"` has been there for months, way before our renaming.

If we allowed events to have name, it would mean the name of the event:

    <event name="Birthday">

This means the event is using a data object called "Birthday"

    <event module="Birthday">

In ASN.X terminology, ASN.1 Data Object's structure/schema is called a module.

e.g. you have 3 data objects about 3 people in the module "person".

Reply:
For posterity: I mean "event" when I typed "name" above, and I was commenting on the wrong snippet. Sorry. James should know what I am referring to now :)